src/java.base/share/classes/jdk/internal/vm/ThreadContainers.java

```diff
@@ -0,0 +1,273 @@
+            if (parent != null)
+                return parent;
+            if ((parent = ThreadContainers.container(owner)) != null)
+                return parent;
+        }
+        ThreadContainer root = ThreadContainers.root();
+        return (container != root) ? root : null;
+    }
+
+    /**
+     * Returns given thread container's "children".
+     */
+    static Stream<ThreadContainer> children(ThreadContainer container) {
+        // children of registered containers
+        Stream<ThreadContainer> s1 = CONTAINER_REGISTRY.stream()
+                .map(WeakReference::get)
+                .filter(c -> c != null && c.parent() == container);
+
+        // container may enclose another container
+        Stream<ThreadContainer> s2 = Stream.empty();
+        if (container.owner() != null) {
+            ThreadContainer next = next(container);
+            if (next != null)
+                s2 = Stream.of(next);
+        }
+
+        // the top-most container owned by the threads in the container
+        Stream<ThreadContainer> s3 = container.threads()
+                .map(t -> Optional.ofNullable(top(t)))
+                .flatMap(Optional::stream);
+
+        return Stream.concat(s1, Stream.concat(s2, s3));
+    }
+
+    /**
+     * Returns the thread container that the given Thread is in or the root
+     * container if not started in a container.
+     */
+    public static ThreadContainer container(Thread thread) {
+        ThreadContainer container = JLA.threadContainer(thread);
+        return (container != null) ? container : root();
+    }
+
+    /**
+     * Returns the top-most thread container owned by the given thread.
+     */
+    private static ThreadContainer top(Thread thread) {
+        StackableScope current = JLA.headStackableScope(thread);
+        ThreadContainer top = null;
+        while (current != null) {
+            if (current instanceof ThreadContainer tc) {
+                top = tc;
+            }
+            current = current.previous();
+        }
+        return top;
+    }
+
+    /**
+     * Returns the thread container that the given thread container encloses.
+     */
+    private static ThreadContainer next(ThreadContainer container) {
+        StackableScope current = JLA.headStackableScope(container.owner());
+        if (current != null) {
+            ThreadContainer next = null;
+            while (current != null) {
+                if (current == container) {
+                    return next;
+                } else if (current instanceof ThreadContainer tc) {
+                    next = tc;
+                }
+                current = current.previous();
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Root container. Contains all platform threads that are not started in a
+     * container, plus some (or all) virtual threads that are started directly
+     * with the Thread API.
+     */
+    private static class RootContainer extends ThreadContainer {
```

Comment:
This implementation could be clearer if split out into two, and the value assigned to `INSTANCE` is selected based  on the system property. Something to consider later perhaps.

Reply:
We've been undecided on whether to just track all threads. If we do that then there would be only one implementation. But yes, if we continue with two then they should be split out.

Reply:
Done.